TrafficReplayer/src/org/opensearch/migrations/replay/RequestResponseResponseTriple.java

```diff
@@ -28,21 +33,54 @@ public RequestResponseResponseTriple(RequestResponsePacketPair sourcePair,
     public static class TripleToFileWriter {
         OutputStream outputStream;
 
-        private JSONObject jsonFromHttpData(List<byte[]> data) {
-            String aggregatedString = new String(Bytes.concat(data.toArray(byte[][]::new)), Charset.defaultCharset());
+        private List<byte[]> extractBodyBytesAfterHeader(List<byte[]> data, int header_length) {
+            // This method essentially iterates through data, discarding the first `header_length` bytes
+            // and then returns the rest in the form of a new List<>, which is a sublist of the input List,
+            // with a potentially modified first element (to discard header bytes).
+            int currently_seen_length = 0;
+            int index = 0;
+            for (byte[] packet : data) {
+                if (currently_seen_length + packet.length < header_length) {
+                    currently_seen_length += packet.length;
+                    index++;
+                    continue;
+                }
+                if (currently_seen_length + packet.length == header_length) {
+                    break;
+                }
+                byte[] new_first_array;
+                if (header_length - currently_seen_length == packet.length) {
+                    new_first_array = packet;
+                } else {
+                    new_first_array = Arrays.copyOfRange(packet, header_length - currently_seen_length, packet.length);
+                }
+                List<byte[]> newList = data.subList(++index, data.size());
+                newList.add(0, new_first_array);
+                return newList;
+            }
+            // If we make it out of this method, the header was the full data, return an empty list.
+            return new ArrayList<byte[]>();
+        }
+
+        private JSONObject jsonFromHttpData(List<byte[]> data) throws IOException {
+
+            SequenceInputStream collatedStream = new SequenceInputStream(Collections.enumeration(data.stream().map(b -> new ByteArrayInputStream(b)).toList()));
```

Comment:
I personally like to take sequences (java.util.Stream) or  in when possible so that you don't have to realize full lists and do work over lots more memory that you aren't going to touch for a while.  Sure, you already have a list coming in - but even repeated intermediary projections at different levels will blow out cache lines.  It's also more painful than it should be to convert streams to enumerations, but you can take a look at [Guava's utility to do just that](https://guava.dev/releases/snapshot/api/docs/com/google/common/collect/Iterators.html#asEnumeration-java.util.Iterator-).